text/0000-stateful-exceptions.md

```diff
@@ -0,0 +1,70 @@
+- Feature Name: stateful-exceptions
+- Start Date: 2017-01-27
+- RFC PR:
+- Pony Issue:
+
+# Summary
+
+Allow exceptions to carry a value from the error site to the handling site, while keeping the exception handling static. The value or its type aren't involved in destination checking. Exceptions will still land at the first handler encountered, with negligible additional runtime cost.
+
+# Motivation
+
+We currently have two very distinct idioms to handle errors in the language and standard library.
+
+1. Exceptions. They are used most of the time but since they are valueless, they can't be used to propagate the reason of an error to a caller function.
+2. Union types of the normal result and the error reason(s). This is used when the reason of an error is needed by a caller function (e.g. the constructors of `File` in the standard library).
```

Comment:
For completeness, the third idiom would be having a notifier object that will be invoked for any errors encountered.